### PR TITLE
solved pp tutorial

### PR DESCRIPTION
**Pull request by ggruszczynski**
Added model and config.xml for pp tutorial.
PR with actual tutorial coming shortly to CFD-GO/TCLB_docs.

### CODE REVIEW

File: models/multiphase/d2q9_pseudo_potential/Dynamics.R [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+# Iteration list
+AddStage("BaseIteration", "Run"     ,  save=Fields$group %in% c("f", "neighbour_type_group") , load=DensityAll$group %in% c("f","neighbour_type_group"))
+AddStage("PsiIteration" , "calcPsi" ,  save=Fields$name=="psi", load=DensityAll$group %in% c("f"))
+
+AddAction("Init"     , c("BaseInit",      "PsiIteration"))
+AddAction("Iteration", c("BaseIteration", "PsiIteration"))
+
+# Output Values
+AddQuantity( name="U",    unit="m/s", vector=TRUE )
+AddQuantity( name="Ueq",  unit="m/s", vector=TRUE )
```
Comment (mdzik) [inline models/multiphase/d2q9_pseudo_potential/Dynamics.R]: Why do you need Ueq here?

File: models/multiphase/d2q9_pseudo_potential/Dynamics.R [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+AddQuantity( name="Rho",  unit="kg/m3" )
+AddQuantity( name="Psi",  unit="1" )
+AddQuantity( name="Neighbour_type",  unit="1" )
+AddQuantity( name="F_ff", unit="N", vector=TRUE)
+AddQuantity( name="F_sf", unit="N", vector=TRUE)
+
+
+# Model Specific Parameters
+AddSetting( name="omega", comment='inverse of relaxation time')
+AddSetting( name="nu", omega='1.0/(3*nu+0.5)', default=0.16666666, comment='viscosity')
```
Comment (mdzik) [inline models/multiphase/d2q9_pseudo_potential/Dynamics.R]: Rename to viscosity

File: models/multiphase/d2q9_pseudo_potential/Dynamics.R [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+# Output Values
+AddQuantity( name="U",    unit="m/s", vector=TRUE )
+AddQuantity( name="Ueq",  unit="m/s", vector=TRUE )
+AddQuantity( name="Rho",  unit="kg/m3" )
+AddQuantity( name="Psi",  unit="1" )
+AddQuantity( name="Neighbour_type",  unit="1" )
+AddQuantity( name="F_ff", unit="N", vector=TRUE)
+AddQuantity( name="F_sf", unit="N", vector=TRUE)
+
```
Comment (mdzik) [inline models/multiphase/d2q9_pseudo_potential/Dynamics.R]: Aren't those values debug-only?

File: models/multiphase/d2q9_pseudo_potential/Dynamics.c.Rt [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+    
+// ------------------------------ pseudo-potential related code --------------------------------------------- //
+
+CudaDeviceFunction void calcPsi() {
+    // Calculate psi at each point so that pseudopotential force can be found.
+    // eq 9.102 p369 from book: 'The Lattice Boltzmann Method: Principles and Practice'
+    // T. Krüger, H. Kusumaatmaja, A. Kuzmin, O. Shardt, G. Silva, E.M. Viggen
+    
+    real_t rho = getRho();
```
Comment (mdzik) [inline models/multiphase/d2q9_pseudo_potential/Dynamics.c.Rt]: This is might be ineffective. Create global variable, set it once and reuse. How many times/collision only you call getRho?

File: models/multiphase/d2q9_ShanChen/Dynamics.R [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+AddDensity( name="f[1]", dx= 1, dy= 0, group="f")
+AddDensity( name="f[2]", dx= 0, dy= 1, group="f")
+AddDensity( name="f[3]", dx=-1, dy= 0, group="f")
+AddDensity( name="f[4]", dx= 0, dy=-1, group="f")
+AddDensity( name="f[5]", dx= 1, dy= 1, group="f")
+AddDensity( name="f[6]", dx=-1, dy= 1, group="f")
+AddDensity( name="f[7]", dx=-1, dy=-1, group="f")
+AddDensity( name="f[8]", dx= 1, dy=-1, group="f")
+
+AddDensity( name="rho", dx=0, dy=0, group="density") # is it the same as  AddField("rho", dx=c(0,0), dy=c(0,0), group="density")
```
Comment (ggruszczynski) [inline models/multiphase/d2q9_ShanChen/Dynamics.R]: Q: Is this the same for all cases?
AddField("xa", dx=c(0,0), dy=c(0,0), group="my_xa") 
// or 
AddDensity( name="xb", dx=0, dy=0, group="my_xb")

then
```c
real_t var1 = xa;
// or
real_t var2 = xa(0,0);

```

File: models/multiphase/d2q9_ShanChen/Dynamics.c.Rt [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+}
+
+
+// ------------------------------ pseudo-potential related code --------------------------------------------- //
+
+CudaDeviceFunction void calcPsi() {
+    // Calculate psi at each point so that pseudopotential force can be found.
+    
+    real_t d = calcRho();
+    // real_t d = getRho(); // this one crashes
```
Comment (ggruszczynski) [inline models/multiphase/d2q9_ShanChen/Dynamics.c.Rt]: I do load "density" during "PsiIteration" but getRho() crashes.

File: models/multiphase/d2q9_pseudo_potential/Dynamics.R (hunk 2 above)
Comment (ggruszczynski) [reply models/multiphase/d2q9_pseudo_potential/Dynamics.R]: changed

File: models/multiphase/d2q9_pseudo_potential/Dynamics.R (hunk 3 above)
Comment (ggruszczynski) [reply models/multiphase/d2q9_pseudo_potential/Dynamics.R]: rather debug, but I think that
```
AddQuantity( name="Rho",  unit="kg/m3" )
AddQuantity( name="Psi",  unit="1" )
AddQuantity( name="F_ff", unit="N", vector=TRUE)
AddQuantity( name="F_sf", unit="N", vector=TRUE)
```
can be keep for tutorial

File: models/multiphase/d2q9_pseudo_potential/Dynamics.c.Rt [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+}
+CudaDeviceFunction real_t getNeighbour_type()
+{
+    return neighbour_type(0,0);
+}
+
+CudaDeviceFunction vector_t getF_ff() {
+    // fluid - fluid interaction force 
+    // eq 9.105 p 372 from book: 'The Lattice Boltzmann Method: Principles and Practice'
+    // T. Krüger, H. Kusumaatmaja, A. Kuzmin, O. Shardt, G. Silva, E.M. Viggen
```
[the remark this answers is not in the document]
Comment (ggruszczynski) [reply models/multiphase/d2q9_pseudo_potential/Dynamics.c.Rt]: done

**Comment by ggruszczynski** [comment]
thx for review @mdzik 
Reiterate please (sorry for mess with renaming) :)

**Comment by llaniewski** [comment]
@mdzik any comments?

File: models/multiphase/d2q9_ShanChen/Dynamics.R (hunk 5 above)
Comment (mdzik) [reply models/multiphase/d2q9_ShanChen/Dynamics.R]: This is not a simple question :)
Densities gets 'advected', are persistent and rw. Fields are "compute-synchronize-allow neighbor access" type of variable. At least as far as I know.

For Shan-Chen gradient evaluation You most likely want Field.
Hope it helps.

File: models/multiphase/d2q9_ShanChen/Dynamics.c.Rt (hunk 6 above)
Comment (mdzik) [reply models/multiphase/d2q9_ShanChen/Dynamics.c.Rt]: I think you might misunderstood me at some point: 
 - each time you call getRho you sum up f's.
- during single stage local class variables (Dynamics.c is an inside of a class) could be used to store some reusable data 
- in S-C type models density/velocity often might be evaluated number of times, even during single stage. there is a room for some time-saving

Do not store density between stages - memory transfer is much more time consuming than sum(f[i])

File: example/multiphase/pseudo_potential/pp.xml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+                        <!-- <Box dx="50" nx="28" dy="50" ny="28"/> -->
+                        <Sphere ny="48" nx="48" dx="40" dy="0" /> 
+                </None>
+        </Geometry>
+        <Model>
+                <Params Density="0.056"/>
+                <Params Density-blobb="2.659"/>
+                <Params G_ff="-6.0"/>
+                <Params G_sf="-3.7"/>
+                <Params nu="0.166"/>
```
Comment (mdzik) [inline example/multiphase/pseudo_potential/pp.xml]: Viscosity

File: models/multiphase/d2q9_ShanChen/Dynamics.R [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+AddAction("Iteration", c("BaseIteration", "PsiIteration"))
+
+# Output Values
+AddQuantity( name="U",    unit="m/s", vector=TRUE )
+AddQuantity( name="Rho",  unit="kg/m3" )
+AddQuantity( name="Psi",  unit="1" )
+AddQuantity( name="Neighbour_type",  unit="1" )
+AddQuantity( name="F_ff", unit="N", vector=TRUE)
+AddQuantity( name="F_sf", unit="N", vector=TRUE)
+
```
Comment (mdzik) [inline models/multiphase/d2q9_ShanChen/Dynamics.R]: I'm quite certain that there is no need to pollute output with those

File: example/multiphase/pseudo_potential/pp.xml [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+                <!-- <Params GravitationY="-1e-5"/> -->
+        </Model>
+        <!-- <Solve Iterations="100" output="output/"> <VTK Iterations="1"/> </Solve>  -->
+        <Solve Iterations="1000" output="output/"> <VTK Iterations="10"/> </Solve> 
+        <Failcheck Iterations="1000" />
```
Comment (mdzik) [inline example/multiphase/pseudo_potential/pp.xml]: Add box to failcheck - no need to evaluate whole domain, especially walls

Comment (ggruszczynski) [reply example/multiphase/pseudo_potential/pp.xml]: So what exactly the empty <Box/> element does?

File: models/multiphase/d2q9_ShanChen/Dynamics.c.Rt (hunk 6 above)
Comment (ggruszczynski) [reply models/multiphase/d2q9_ShanChen/Dynamics.c.Rt]: ok, I can agree that some simple functions like sum(f[i]) can be sometimes quicker than loading memory.
However, I still dont understand why calcRho() which sums up 'f' works while getRho() crashes.
AFAIK getRho() shall read the rho from previous stage... or maybe it reads it from previous iteration?

File: models/multiphase/d2q9_ShanChen/Dynamics.R (hunk 5 above)
Comment (ggruszczynski) [reply models/multiphase/d2q9_ShanChen/Dynamics.R]: by 'advected' means that one can write to other density cell?
for example 
my_density[1] = 123; // ok
my_field(0,1) = 123; // shall crash?

however
real_t var1 = myfield; // ok
real_t var2 = myfield(0,0); // ok - same as var1

myfield(0,0) =123; // ok? or is it an read access pattern only?
myfield = 234; //ok

File: models/multiphase/d2q9_ShanChen/Dynamics.c.Rt (hunk 6 above)
Comment (ggruszczynski) [reply models/multiphase/d2q9_ShanChen/Dynamics.c.Rt]: Into which class Dynamics.c gets injected?

File: models/multiphase/d2q9_ShanChen/Dynamics.R (hunk 5 above)
Comment (ggruszczynski) [reply models/multiphase/d2q9_ShanChen/Dynamics.R]: While having `AddDensity( name="rho", dx=0, dy=0, group="density")`
```c
CudaDeviceFunction real_t getRho() {
    // This function defines the macroscopic density at the current node.
    return rho; // works fine, and Rho is in the output
}
```

File: example/multiphase/pseudo_potential/pp.xml (hunk 10 above)
Comment (llaniewski) [reply example/multiphase/pseudo_potential/pp.xml]: I think @mdzik is refering to this: https://github.com/CFD-GO/TCLB/blob/develop/src/Handlers/cbFailcheck.cpp.Rt#L18
You can specify a region to check like this: `<Failcheck ... dx="10" nx="10"/>`

On the other hand, I would not consider it crusial.

**Comment by ggruszczynski** [comment]
...

**Comment by llaniewski** [comment]
Reopening.

**Pull request merged**